### PR TITLE
Keep ordering for parsing cookies as it is important

### LINKED ISSUE #25277: cookie parsing bug
### Describe the bug

Because there is no `if branch` to match `SameSite`(or others) and **the cookie name (or value) is allowed to be overridden**, the cookie name (or value) is incorrectly set to `SameSite`(or others)

```
auth_key=123456; SameSite=None
 --> parsing to -->
cookie.name=SameSite, cookie.value=None 
```

@see https://github.dev/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/headers/NewCookieHeaderDelegate.java#L21

Web capture_30-4-2022_94152_developer mozilla org

### Expected behavior

```
auth_key=123456; SameSite=None
 --> parsing to -->
cookie.name=auth_key, cookie.value=123456
```

### Actual behavior

```
auth_key=123456; SameSite=None
 --> parsing to -->
cookie.name=SameSite, cookie.value=None 
```

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.8.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by ejba**
Fixes #25277

### CODE REVIEW

File: integration-tests/resteasy-reactive/pom.xml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    <parent>
+        <artifactId>quarkus-integration-tests-parent</artifactId>
+        <groupId>io.quarkus</groupId>
+        <version>999-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>quarkus-integration-test-resteasy-reactive</artifactId>
+
+    <name>Quarkus - Integration Tests - RESTEasy Reactive</name>
```
Comment (gsmet) [inline integration-tests/resteasy-reactive/pom.xml:15]: I think this needs to be included in the RESTEasy Reactive tests. See https://github.com/quarkusio/quarkus/pull/25291/files for an example.

**Comment by gsmet** [comment]
I agree it's probably better to do it this way but... AFAICS a RESTEasy Reactive TCK test is not passing.

/cc @geoand

**Comment by geoand** [comment]
If the TCK isn't passing, we can't include the fix unfortunately

**Comment by ejba** [comment]
I agree. I'm trying to understand what the failed test does to fix this.

If it's taking too long, I'll add a new branch to ignore "SameSite" (requires less changes).

**Comment by ejba** [comment]
Ok, so the problem is what is the default cookie's version when it's not specified in the `set-cookie`.

I had a look into the test, and I found something odd on it. The test description mentions `Create a version 0`, however it expects a cookie with version 1. 

```    
    /*
     * @testName: parseTest1
     * 
     * @assertion_ids: JAXRS:JAVADOC:111; JAXRS:JAVADOC:100; JAXRS:JAVADOC:101;
     * JAXRS:JAVADOC:103; JAXRS:JAVADOC:53; JAXRS:JAVADOC:54; JAXRS:JAVADOC:55;
     * JAXRS:JAVADOC:56;
     * 
     * @test_Strategy: Create a version 0 Cookie instance by Parsing a String <------
     */
    @org.junit.jupiter.api.Test
    public void parseTest1() throws Fault {
        String NewCookie_toParse = "NAME_1=Value_1;";
        String name = "name_1";
        String value = "value_1";
        String path = "";
        String domain = "";
        int version = 1;

        NewCookie nck26 = javax.ws.rs.core.NewCookie.valueOf(NewCookie_toParse);

        verifyNewCookie(nck26, name, value, path, domain, version, "", -1, false);
    }
  ```
  
So I decided to look into the RFCs to confirm what is the default cookie's version when it's not specified, and I got these rules: 

>    The value of the cookie-version attribute must be the value from the
   Version attribute, if any, of the corresponding Set-Cookie response
   header.  Otherwise the value for cookie-version is 0.

(source)[https://www.ietf.org/rfc/rfc2109.txt]


>    The value of the cookie-version attribute MUST be the value from the
   Version attribute of the corresponding Set-Cookie2 response header.
   Otherwise the value for cookie-version is 0.  

(source)[https://www.ietf.org/rfc/rfc2965.txt]

**Comment by gsmet** [comment]
Probably worth checking that we haven’t changed that ourselves.

**Comment by ejba** [comment]
`HttpCookie.parse` follows the RFC specification, so creates version 0 cookies when not specified. I don't mind reversing this approach and adding the branch to "SameSite", I just found it strange that even the tck test document refers that it will test a version 0 cookie creation when testing a version 1.

**Comment by gsmet** [comment]
So, for me, we have two options:
1. use `HttpCookie` and force the version to 1 if it is 0. Basically `HttpCookie` is doing smart things to detect if it's using old cookies or new. That's why it sets the version to 0. Problem is: no idea if it will change the behavior in some other cases.
2. fix our code and it's not only about adding `SameSite`: we need to change the logic to:
  - use a `LinkedHashMap` in `ParameterParser` - or use a specific parser for cookies
  - once we have done that: the first item is the name/value of the cookie, the rest are the options of the cookie

RESTEasy Classic is doing 2/. Apparently, we have deliberately broken things in RESTEasy Reactive by optimizing `ParameterParser` to not return a `LinkedHashMap` and the cookie parsing to not rely on the ordering.
This is incorrect.

If we want to keep the optimization in the general case, we need a specific `CookieParser` that keeps the ordering.

**Comment by gsmet** [comment]
@ejba @geoand this is what I have in mind: https://github.com/gsmet/quarkus/tree/cookie-parser

**Comment by ejba** [comment]
@gsmet I implemented the parser based on the HttpCookie algorithm and with some adaptations for the current use case.

File: integration-tests/resteasy-reactive/src/main/java/io/quarkus/it/resteasy/server/GreetingResource.java [hunk 2]
```diff
@@ -10,7 +10,7 @@ public class GreetingResource {
 
     @GET
     public Response hello() {
-        return Response.status(200).cookie(NewCookie.valueOf("Set-Cookie: greeting=hello; SameSite=Lax")).build();
+        return Response.status(200).cookie(NewCookie.valueOf("Set-Cookie: greeting=hello; SameSite=Lax;")).build();
```
Comment (gsmet) [inline integration-tests/resteasy-reactive/src/main/java/io/quarkus/it/resteasy/server/GreetingResource.java:13]: Hmmm. This worries me a bit. You don't trigger the issue if you don't have a trailing `;`?

**Comment by ejba** [comment]
Sorry @gsmet , I should mentioned first I was working on this, my bad. 
I don't mind if you prefer revert my changes.

**Comment by gsmet** [comment]
@ejba I think I would rather be more cautious and keep the approach that we were using in RESTEasy (see my comment above) rather than pushing a brand new implementation.
Let's discuss it with @geoand tomorrow, today is a public holiday in Greece.

**Comment by ejba** [comment]
np, I understand!

File: integration-tests/resteasy-reactive/src/main/java/io/quarkus/it/resteasy/server/GreetingResource.java (hunk 2 above)
Comment (ejba) [reply integration-tests/resteasy-reactive/src/main/java/io/quarkus/it/resteasy/server/GreetingResource.java]: I've included more tests to check that the header doesn't have a trailing `;`, this won't raise any issues.

File: independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/headers/CookieParser.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+        return parseInternal(header);
+    }
+
+    private static boolean startsWithIgnoreCase(String s, String start) {
+        if (s == null || start == null)
+            return false;
+
+        if (s.length() >= start.length() &&
+                start.equalsIgnoreCase(s.substring(0, start.length()))) {
```
Comment (gastaldi) [inline independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/headers/CookieParser.java:40]: Can be simplified to: `s.regionMatches(true, 0, start, 0, start.length())`

Comment (gsmet) [reply independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/headers/CookieParser.java:40]: This is coming from the JDK so if we go this route, I would keep the logic as equivalent as possible.

Anyway, let's decide on a direction first before entering into the details.

**Comment by geoand** [comment]
> @ejba @geoand this is what I have in mind: https://github.com/gsmet/quarkus/tree/cookie-parser and more specifically [gsmet@0425b34](https://github.com/gsmet/quarkus/commit/0425b34e4279ce3d6555363d252c0517b556f591)

I think this make most sense

**Review by geoand — APPROVED** [review]
Thanks!

**Comment by ejba** [comment]
@geoand Can I mark this PR as "Ready for Review"?

**Comment by geoand** [comment]
I did for you :)

**Review by gsmet — CHANGES_REQUESTED** [review]
Please adjust the commit comment. You kept the old one and it's misleading. Thanks!

**Comment by ejba** [comment]
@gsmet I re-used your commit message, hope you don't mind.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Thanks @ejba ! Nice team work.

**Pull request merged**